ASoC: SOF: Intel: hda-loader: Correct cl_dsp_init sequence

ujfalusi opened the pull request:
The first step of the power on sequence should not unstall and run the
core(s).
The unstall/run step must be done after we prepared the ROM_CONTROL
message for the hardware to purge or do an IMR boot.

The ROM_CONTROL message must be set to the IPC register before the core
unstall/run.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
While I'm a  bit nervous about this patch since it is changing the DSP boot flow for both cold/purge and IMR boot, I have spent two days on testing it on different machines and especially on ones which _known_ to produce IMR boot failures.
I have boards in IMR boot loop (aplay -d5; sleep 5;) and so far they have completed 2095 (TGL), 2090 (APL), 2367 (TGLU) and 410 (JSL) cycles without a single boot issue.
On JSL it is running rmmod/insmod loop for the 70th time and no issue.
The PR check looks good also all over the boards.

kv2019i reviewed (approved):
Excellent catch @ujfalusi ! To help maintainers put this into context, it would be useful to describe in the commit message whether this is a recent regression, or only affects IMR. If I understand correctly, this is fixing the "skip-purge-IMR-boot" flow and does not impact earlier versions. If that's a case, a fixes-tag would help maintainers backport this to correct kernels.

ujfalusi commented:
@kv2019i, this boot flow is in place since the beginning (the untall/run twice). We always had random and unexplained boot failures which might be caused by this, but I would not dare to claim it.
In cold boot we have much more time between the unstall/run and the firmware booting up due to the firmware download from host.
We started to see IMR issues when I switch the IMR flow to re-use the common boot sequence.
I think what happened is that depending on timing the ROM_CONTROL message got written _after_ the cores are running, the ROM did not take it out and the booted firmware tries to handle it. This is really random and depends on scheduling delays and likely HW signaling also.

We can not put fixes tag as I'm sure it can not be backported to kernels from 2018, but it should accompanies the common boot flow patch.

plbossart commented on sound/soc/sof/intel/hda-loader-skl.c:
```diff
@@ -341,7 +341,7 @@ static int cl_dsp_init_skl(struct snd_sof_dev *sdev,
 			goto err;
 		}
-		ret = hda_dsp_enable_core(sdev, chip->init_core_mask);
+		ret = hda_dsp_enable_core(sdev, chip->init_core_mask, true);
```
we need to find a way to split this SKL part from the rest if we want to upstream this code, otherwise this fix will be gated by the SKL patches which could take a while.

I also don't quite get why we use 'true' here and 'false' for more recent cavs platforms - it's also possible that I misunderstood the flows since a different DMA is used.

ujfalusi replied:
@plbossart, I see. Then to avoid compile errors, I'll re-do this completely with new function names to use.

Since I don't have access to SKL device I can not test it and I could not find documentation either. To avoid any change in code I can not test, I left the sequence there as it was.

ranj063 replied:
@ujfalusi I have a KBL laptop. I can test it for you once you have updated it

plbossart replied:
I am the only one with a SKL device, I can test - should be the same as KBL but let's see. 
There's also KBL devices in the Intel CI pool, you can reserve them at will @ujfalusi

ujfalusi replied:
I think the sequence for skl is correct as far as I can tell. There is no ROM_CONTROL message, no IMR boot is supported, so no need to purge the storage?

Ok, I'll reserve a KBL device for testing, but as far as I can tell this looks OK as it is.

ranj063 commented:
@ujfalusi 3 questions:
1. you commit message says we should unstall/run after ROM_CONTROL message, so where are we doing it?
2. Can we please align with HW team if this sequence is indeed correct?
3. With this change, can we remove the change that keyon added to clear done bit?

ujfalusi commented:
> @ujfalusi 3 questions:
> 
>     1. you commit message says we should unstall/run after ROM_CONTROL message, so where are we doing it?

https://github.com/thesofproject/linux/blob/db34128c2a8d8456ddf6fd5ac18fda7e85976d62/sound/soc/sof/intel/hda-loader.c#L130

>     2. Can we please align with HW team if this sequence is indeed correct?

Sure, I have checked in several documents and all told the same sequence
 
>     3. With this change, can we remove the change that keyon added to clear done bit?

I'm not sure about that, it needs to be tested, it looked like that in APL it is needed.

ujfalusi commented:
@ranj063, the problem, as far as I see it is that currently the `hda_dsp_enable_core()` includes `hda_dsp_core_run()` right after the powe on.
So we unstall/run the cores _before_ writing the ROM_CONTROL message to IPC register and this is against the documented flows. Then we unstall/run again and I'm not sure what that does if anything on an unstalled/running core.

I'll update the PR to leave the `hda_dsp_enable_core()` alone and add a new (gash) function to only do the power on (we need similar checks in there, so can not nest them), let's see how that will look as I don't have the code on front of me atm

ujfalusi commented:
Changes since v1:
- expose and use hda_dsp_core_power_up() instead of changing the the hda_dsp_enable_core() parameter list.
 - no changes outside of hda-loader.c (and hda-dsp.c hda.h)
 - the already stress tested flow is preserved

ujfalusi commented:
@plbossart, @ranj063, @keqiaozhang, I can reserve the JSL and CML device from CI and run this PR on them for an extended time if that would help to build up confidence on this change.

btw: while looking at the `cl_dsp_init()` it made me wonder why we power on all host managed cores, then unstall/run core0 (!) only. Finally we power down all cores which are not init cores (init core is core0 in all cases.
Why we need to power on all cores just to power them down and in the documentation only core0 is mentioned for code loading.
This is a subject of another PR anyways, just interesting.

ujfalusi commented:
@ranj063, initially we had `snd_sof_dsp_core_power_up()` as first step in `cl_dsp_init()` but it got replaced by `hda_dsp_enable_core()` which does both power_up and unstall/run by this commit:

f8c2d22211144 ("ASoC: SOF: hda: don't use the core op for power up/power down")

PS: ~1100 IMR boot iteration is completed on the JSL and CML device from CI. ~500 on TGL and APL locally.

ujfalusi commented:
Finished the testing: JSL, CML over 1800 IMR boot each, TGL, APL over 1200 IMR boot each. Can not reproduce the issue with this PR.

ranj063 commented on sound/soc/sof/intel/hda-dsp.c:
```diff
@@ -182,12 +182,20 @@ int hda_dsp_core_run(struct snd_sof_dev *sdev, unsigned int core_mask)
  */
 
-static int hda_dsp_core_power_up(struct snd_sof_dev *sdev, unsigned int core_mask)
+int hda_dsp_core_power_up(struct snd_sof_dev *sdev, unsigned int core_mask)
```
typo in commit messgae "correc"

ranj063 reviewed (approved):
looks good except for a minor typo

ujfalusi commented:
Changes since v2:
- correct type in commit message for the first patch

ujfalusi commented:
Fwiw, out of curiosity I re-run the same test on the JSL and CML device without this PR.
JSL reproduced the `no reply expected, received 0x0, will be ignored` within 500 iterations, CNL is still running with no issue.
In my previous tests on JSL I was able to reproduce the problem under 600 consistently without this PR, so surviving 1800 is a good indication.

plbossart commented:
@ujfalusi JSL is a bit of a corner case. It's cAVS 2.1 (based on ICL), but with 2 cores so we can't use the ICCMAX work-around with core3 that's used on ICL, and may also be required on TGL.

We should ideally run this test on ICL and TGL to make sure we're not missing something based on a small test sample.

ujfalusi commented:
@plbossart, I have run this on tgl locally and on one CI machine, passing over 1000 IMR boot. I would be more confident after 10k runs, but combined all the runs, I'm over that.
with this PR I have not aeen a single fail so far

ujfalusi commented:
@plbossart, I have done 2000 IMR boots on two machines from CI (ICL and TGLU), locally I'm at 1500 IMR boot on TGL, TGL-H and and APL machine, not observed a single case so far.

ujfalusi commented:
Changes since v3:
- new patch added to clarify, update the comment for `cl_dsp_init()` for clarity and use the chip->init_core_mask instead of BIT(0) to unstall/run the boot/init core.

ujfalusi commented:
Had to stop the testing on the local TGL, TGL-H and APL machines after 2100 IMR boot iteration, no issue found. Altogether more than 10K IMR boots ;)

plbossart commented:
@ranj063 @kv2019i can you review?
